src/richie/apps/search/forms.py

```diff
@@ -167,18 +167,45 @@ def get_queries(self):
                     "key": "query",
                     "fragment": [
                         {
-                            "multi_match": {
-                                "analyzer": QUERY_ANALYZERS[lang],
-                                "fields": [
-                                    "description.*",
-                                    "introduction.*",
-                                    "title.*",
-                                    f"categories_names.*^{related_content_boost}",
-                                    f"organizations_names.*^{related_content_boost}",
-                                    f"persons_names.*^{related_content_boost}",
-                                ],
-                                "query": full_text,
-                                "type": "best_fields",
+                            "bool": {
+                                "should": [
+                                    {
+                                        "multi_match": {
+                                            "analyzer": QUERY_ANALYZERS[lang],
+                                            "fields": [
+                                                "description.*",
+                                                "introduction.*",
+                                                "title.*",
+                                                f"categories_names.*^{related_content_boost}",
+                                                f"organizations_names.*^{related_content_boost}",
+                                                f"persons_names.*^{related_content_boost}",
+                                            ],
+                                            "query": full_text,
+                                            "type": "best_fields",
+                                        }
+                                    },
+                                    # Boost exact code matches so that they jump our state buckets
+                                    # A closed course with a perfect code match should be
+                                    # displayed bedore an open course with no perfect code match
+                                    {
+                                        "match": {
+                                            "code": {
+                                                "query": full_text,
+                                                "analyzer": "code",
+                                                "boost": 20,
```

Comment:
I tested it locally to search a closed course through its exact code and open courses with no perfect match are still displayed before.

<img width="982" alt="Capture d’écran 2021-08-09 à 10 23 13" src="https://user-images.githubusercontent.com/9265241/128677977-535dfdd8-3ee1-4d8c-b9da-615d4f970dcf.png">
<img width="983" alt="Capture d’écran 2021-08-09 à 10 23 04" src="https://user-images.githubusercontent.com/9265241/128677983-585c0a14-9e81-40bf-88a7-7e43b18e80f6.png">

Reply:
True it seems to be because of fuzziness which I added later. Let me write a test for this.

Reply:
@jbpenrath I added a test and adjusted the field boost. Can you test again?

Reply:
👍  Great ! It works as expected. Search `00046` returns the course with code `00046` first.
Search `004` returns all opened course containing the trigram `004` first then other course containing this trigram and finally other courses.

Reply:
cool